### PR TITLE
Feature/application service refactoring

### PR DESCRIPTION
**Pull request by MariusWirtz**
New `Application` class and dedicated subclass for all application types
(`CubeApplication`, `DocumentApplication`, etc.) contains specific logic for json creation and interpretation.

`ApplicationService` now has `get`, `get_document`, `delete`, `create`,
`exists`, `create_document_from_file` methods.

Add test cases for applications

### CODE REVIEW

File: TM1py/Services/ApplicationService.py [hunk 1]
```diff
@@ -15,143 +16,161 @@ def __init__(self, tm1_rest):
...
+
+        elif application_type == ApplicationTypes.VIEW:
+            response = self._rest.GET(base_url + "?$expand=View($select=Name;$expand=Cube($select=Name))")
+            return ViewApplication(
+                path=path,
+                name=name,
+                cube_name=response.json()["View"]["Cube"]["Name"],
+                view_name=response.json()["View"]["Name"])
+
+    def get_document(self, path, name, private=False):
```
Comment (tombrzy) [inline TM1py/Services/ApplicationService.py]: Let's make it even more consistent with the rest of the TM1py code.
All 'old' methods, for example, the get or update in SubsetService assuming the private parameter as optional and True as default. I suggest to synchronize it and make it consistent. Either everywhere 'private=False' (my favorite :) ) or the other way around.

File: TM1py/Services/ApplicationService.py [hunk 2]
```diff
@@ -15,143 +16,161 @@ def __init__(self, tm1_rest):
...
+            contents=contents,
+            name=name)
         response = self._rest.GET(request)
-        content = response.content
-
-        return Application(path, content)
+        return DocumentApplication(path, name, response.content)
 
-    def delete(self, path, application_type, application_name, access='public'):
+    def delete(self, path, application_type, application_name, private=False):
```
Comment (tombrzy) [inline TM1py/Services/ApplicationService.py]: Inconsistent 'private' parameter default value.

File: TM1py/Services/ApplicationService.py [hunk 3]
```diff
@@ -15,143 +16,161 @@ def __init__(self, tm1_rest):
...
+            contents=contents,
+            application_name=application_name)
 
-        response = self._rest.DELETE(request)
-        return response   
+        return self._rest.DELETE(request)
 
-    def create(self, path, application_type, application_name, object_reference_list='None', access_type='public'):
-        """ Create Planning Analytics application process reference
+    def create(self, application, private=False):
```
Comment (tombrzy) [inline TM1py/Services/ApplicationService.py]: Inconsistent 'private' parameter default value.

File: TM1py/Services/ApplicationService.py [hunk 4]
```diff
@@ -15,143 +16,161 @@ def __init__(self, tm1_rest):
...
+            request = "/api/v1/Contents('Applications'){}')/{}('{}.blob')/Document/Content".format(
+                mid[:-2],
+                contents,
+                application.name)
+            response = self._rest.PUT(request, application.content)
 
         return response
 
-
+    def exists(self, path, application_type, name, private=False):
```
Comment (tombrzy) [inline TM1py/Services/ApplicationService.py]: Inconsistent 'private' parameter default value.

File: TM1py/Services/ApplicationService.py [hunk 5]
```diff
@@ -15,143 +16,161 @@ def __init__(self, tm1_rest):
         """
         self._rest = tm1_rest
 
-    def get(self, path):
+    def get(self, path, application_type, name, private=False):
```
Comment (tombrzy) [inline TM1py/Services/ApplicationService.py]: Inconsistent 'private' parameter default value.

File: TM1py/Services/ApplicationService.py (hunk 1 above)
Comment (MariusWirtz) [reply TM1py/Services/ApplicationService.py]: I like it.
Will change the other `private` default values to `False`

File: TM1py/Services/RESTService.py [hunk 6]
```diff
@@ -148,40 +148,82 @@ def __exit__(self, exception_type, exception_value, traceback):
...
+        :param timeout: Number of seconds that the client will wait to receive the first byte.
+        :return: response object
         """
-        return self._s.patch(url=request, headers=self._headers, data=data, verify=self._verify, timeout=self._timeout)
+        return self._s.patch(
+            url=request,
+            headers={**self._headers, **headers} if headers else self._headers,
+            data=data,
+            verify=self._verify,
+            timeout=timeout if timeout else self._timeout)
```
Comment (MariusWirtz) [inline TM1py/Services/RESTService.py]: Hi @pbuncik,

please check if this implementation of setting custom headers on individual requests also serves your purpose.

**Pull request merged**